File: apps/actix-service/src/controllers/user_controller.rs
use actix_web::{Responder, web};
use serde_json::json;
use crate::helper::base_response::BaseResponse;
use crate::services::user_service::UserService;

/// Get all users
pub async fn get_users() -> impl Responder {
    let user_service = UserService::new();
    
    match user_service.get_users().await {
        Ok(users) => BaseResponse::send_response("success", "Users retrieved successfully", Some(json!(users))),
        Err(error) => BaseResponse::send_response("internalServerError", &format!("Failed to retrieve users: {}", error), None)
    }
}

/// Get a user by ID
pub async fn get_user_by_id(path: web::Path<u32>) -> impl Responder {
    let user_id = path.into_inner();
    let user_service = UserService::new();
    
    match user_service.get_user_by_id(user_id).await {
        Ok(Some(user)) => BaseResponse::send_response("success", "User retrieved successfully", Some(user)),
        Ok(None) => BaseResponse::send_response("unauthorized", "User not found", None),
        Err(error) => BaseResponse::send_response("internalServerError", &format!("Failed to retrieve user: {}", error), None)
    }
}

/// Create a new user
pub async fn create_user(user_data: web::Json<serde_json::Value>) -> impl Responder {
    let user_service = UserService::new();
    
    match user_service.create_user(user_data.into_inner()).await {
        Ok(user) => BaseResponse::send_response("created", "User created successfully", Some(user)),
        Err(error) => BaseResponse::send_response("internalServerError", &format!("Failed to create user: {}", error), None)
    }
}

/// Update an existing user
pub async fn update_user(path: web::Path<u32>, user_data: web::Json<serde_json::Value>) -> impl Responder {
    let user_id = path.into_inner();
    let user_service = UserService::new();
    
    match user_service.update_user(user_id, user_data.into_inner()).await {
        Ok(Some(user)) => BaseResponse::send_response("success", "User updated successfully", Some(user)),
        Ok(None) => BaseResponse::send_response("unauthorized", "User not found", None),
        Err(error) => BaseResponse::send_response("internalServerError", &format!("Failed to update user: {}", error), None)
    }
}

/// Delete a user
pub async fn delete_user(path: web::Path<u32>) -> impl Responder {
    let user_id = path.into_inner();
    let user_service = UserService::new();
    
    match user_service.delete_user(user_id).await {
        Ok(true) => BaseResponse::send_response("success", "User deleted successfully", None),
        Ok(false) => BaseResponse::send_response("unauthorized", "User not found", None),
        Err(error) => BaseResponse::send_response("internalServerError", &format!("Failed to delete user: {}", error), None)
    }
}

File: apps/vertx-service/src/main/java/com/myapp/controller/UserController.java
package com.myapp.controller;

import com.myapp.helper.BaseResponse;
import com.myapp.service.UserService;
import io.vertx.ext.web.RoutingContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;
import java.util.Optional;

public class UserController {

    private static final Logger logger = LoggerFactory.getLogger(UserController.class);

    public static void Index(RoutingContext context) {
        logger.info("Handling Index request");
        BaseResponse.sendResponse(context, "Welcome to Vertx Service", "success", null);
    }

    // Get all users
    public static void getUsers(RoutingContext context) {
        logger.info("User Controller | getUsers");

        try {
            List<Map<String, Object>> users = UserService.getAllUsers();
            logger.info("Returning {} users", users.size());
            BaseResponse.sendResponse(context, "Request successful", "success", users);
        } catch (Exception e) {
            logger.error("Error getting users", e);
            BaseResponse.sendResponse(context, "Internal server error", "error", null);
        }
    }

    // Get user by ID
    public static void getUserById(RoutingContext context) {
        String userId = context.pathParam("id");
        logger.info("User Controller | getUserById for ID: {}", userId);

        if (userId == null || userId.isEmpty()) {
            logger.warn("User ID is required, returning error");
            BaseResponse.sendResponse(context, "User ID is required", "forbidden", null);
            return;
        }

        try {
            Optional<Map<String, Object>> user = UserService.getUserById(userId);
            if (user.isPresent()) {
                logger.info("User found: {}", user.get());
                BaseResponse.sendResponse(context, "Request successful", "success", user.get());
            } else {
                logger.warn("User not found with ID: {}", userId);
                BaseResponse.sendResponse(context, "User not found", "not_found", null);
            }
        } catch (Exception e) {
            logger.error("Error getting user by ID: {}", userId, e);
            BaseResponse.sendResponse(context, "Internal server error", "error", null);
        }
    }

    // Create new user
    public static void createUser(RoutingContext context) {
        logger.info("User Controller | createUser");

        try {
            // Get user data from request body
            Map<String, Object> userData =  context.body().asJsonObject().getMap();
            
            if (userData == null || userData.isEmpty()) {
                logger.warn("User data is required");
                BaseResponse.sendResponse(context, "User data is required", "bad_request", null);
                return;
            }

            Map<String, Object> newUser = UserService.createUser(userData);
            logger.info("User created successfully: {}", newUser);
            BaseResponse.sendResponse(context, "User created successfully", "success", newUser);
            
        } catch (Exception e) {
            logger.error("Error creating user", e);
            BaseResponse.sendResponse(context, "Internal server error", "error", null);
        }
    }

    // Update user by ID
    public static void updateUser(RoutingContext context) {
        String userId = context.pathParam("id");
        logger.info("User Controller | updateUser for ID: {}", userId);

        if (userId == null || userId.isEmpty()) {
            logger.warn("User ID is required, returning error");
            BaseResponse.sendResponse(context, "User ID is required", "forbidden", null);
            return;
        }

        try {
            // Get user data from request body
            Map<String, Object> userData = context.body().asJsonObject().getMap();
            
            if (userData == null || userData.isEmpty()) {
                logger.warn("User data is required");
                BaseResponse.sendResponse(context, "User data is required", "bad_request", null);
                return;
            }

            Optional<Map<String, Object>> updatedUser = UserService.updateUser(userId, userData);
            if (updatedUser.isPresent()) {
                logger.info("User updated successfully: {}", updatedUser.get());
                BaseResponse.sendResponse(context, "User updated successfully", "success", updatedUser.get());
            } else {
                logger.warn("Failed to update user with ID: {}", userId);
                BaseResponse.sendResponse(context, "Failed to update user", "error", null);
            }
        } catch (Exception e) {
            logger.error("Error updating user by ID: {}", userId, e);
            BaseResponse.sendResponse(context, "Internal server error", "error", null);
        }
    }

    // Delete user by ID
    public static void deleteUserById(RoutingContext context) {
        String userId = context.pathParam("id");
        logger.info("User Controller | deleteUserById for ID: {}", userId);

        if (userId == null || userId.isEmpty()) {
            logger.warn("User ID is required, returning error");
            BaseResponse.sendResponse(context, "User ID is required", "forbidden", null);
            return;
        }

        try {
            Optional<Map<String, String>> result = UserService.deleteUserById(userId);
            if (result.isPresent()) {
                logger.info("User deleted successfully: {}", result.get());
                BaseResponse.sendResponse(context, "User deleted successfully", "success", result.get());
            } else {
                logger.warn("Failed to delete user with ID: {}", userId);
                BaseResponse.sendResponse(context, "Failed to delete user", "error", null);
            }
        } catch (Exception e) {
            logger.error("Error deleting user by ID: {}", userId, e);
            BaseResponse.sendResponse(context, "Internal server error", "error", null);
        }
    }
}
